src/main/java/org/jabref/logic/importer/FulltextFetchers.java

```diff
@@ -36,26 +48,93 @@ public FulltextFetchers(ImportFormatPreferences importFormatPreferences) {
         Optional<DOI> doi = clonedEntry.getField(FieldName.DOI).flatMap(DOI::parse);
 
         if (!doi.isPresent()) {
+            findDoiForEntry(clonedEntry);
+        }
+
+        ExecutorService executor = Executors.newCachedThreadPool();
+        try {
+            List<Future<Optional<FetcherResult>>> result = new ArrayList<>();
             try {
-                WebFetchers.getIdFetcherForIdentifier(DOI.class)
-                        .findIdentifier(clonedEntry)
-                        .ifPresent(e -> clonedEntry.setField(FieldName.DOI, e.getDOI()));
-            } catch (FetcherException e) {
-                LOGGER.debug("Failed to find DOI", e);
+                result = executor.invokeAll(getCallables(clonedEntry, finders), FETCHER_TIMEOUT, TimeUnit.SECONDS);
+            } catch (InterruptedException ignore) {
+
             }
+
+            return result.stream()
+                    .map(FulltextFetchers::waitForResult)
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .filter(res -> Objects.nonNull(res.source))
+                    .sorted(Comparator.comparingInt((FetcherResult res) -> res.trust.getTrustScore()).reversed())
+                    .map(res -> res.source)
+                    .findFirst();
+        } finally {
+            shutdownAndAwaitTermination(executor);
         }
+    }
 
-        for (FulltextFetcher finder : finders) {
+    private void findDoiForEntry(BibEntry clonedEntry) {
+        try {
+            WebFetchers.getIdFetcherForIdentifier(DOI.class)
+                    .findIdentifier(clonedEntry)
+                    .ifPresent(e -> clonedEntry.setField(FieldName.DOI, e.getDOI()));
+        } catch (FetcherException e) {
+            LOGGER.debug("Failed to find DOI", e);
+        }
+    }
+
+    private static Optional<FetcherResult> waitForResult(Future<Optional<FetcherResult>> future) {
+        while (true) {
             try {
-                Optional<URL> result = finder.findFullText(clonedEntry);
+                return future.get();
+            } catch (InterruptedException retry) {
+
+            } catch (ExecutionException | CancellationException e) {
+                LOGGER.debug("Fetcher execution failed or was cancelled");
+                return Optional.empty();
+            }
+        }
+    }
+
+    private Callable<Optional<FetcherResult>> getCallable(BibEntry entry, FulltextFetcher fetcher) {
+            return () -> {
+                try {
+                    Optional<URL> result = fetcher.findFullText(entry);
+
+                    if (result.isPresent() && new URLDownload(result.get().toString()).isPdf()) {
+                        return Optional.of(new FetcherResult(fetcher.getTrustLevel(), result.get()));
```

Comment:
Using `Optional.map` may be a bit easier to understand.

Reply:
What kind of code should replace this?

Reply:
I was thinking about something like
````java
return fetcher.findFullText(entry)
          .filter(isPdf())
          .map(url -> new FetcherResult(fetcher.getTrust(), url));
````